### PR TITLE
Bugbyte 143 moderating services

### PR DESCRIPTION
**Pull request by habibatarek26**
**Title**  
Moderator and Admin Services  

**Description**  
The following Moderator Services have been implemented:
- Remove Member
- Remove Post  

Admin Services (in addition to Moderator Services) now include:
- Delete Community
- Remove Moderator
- Set Moderator  

**Changes**  
- Modifications were made to the Community Controller, with related tests resolved by @RowanYoussef.
- New controllers for admin and moderator services have been added.
- Tests were updated and added in the following classes:
  - `CommunityRepositoryTest`
  - `CommunityServiceTest`
  - `CommunityAdapter`
  - `AdministrativeFacadeTest`

### CODE REVIEW

File: BugByte-backend/src/main/java/com/example/BugByte_backend/facades/AdministrativeFacade.java [hunk 1]
```diff
@@ -199,8 +199,10 @@ public Map<String, Object> googleOAuthSignUp(@RequestBody Map<String, String> re
     public Community getCommunityInfo(Map<String,Object> map) throws Exception {
         String token = (String) map.get("jwt");
         String userName = authenticationService.getUserNameFromJwt(token);
+        System.out.println("username ="+userName);
         if (userName == null) throw new Exception("userName is null");
         Community community = communityService.getCommunityById((Long)map.get("communityId"));
+        System.out.println("community "+community.toString());
         System.out.println(map.get("communityId"));
         CommunityAdapter communityAdapter = new CommunityAdapter();
         return community;//communityAdapter.toMap(community);
```
Comment (RowanYoussef) [inline BugByte-backend/src/main/java/com/example/BugByte_backend/facades/AdministrativeFacade.java:208]: unnecessary prints

File: BugByte-backend/src/main/java/com/example/BugByte_backend/facades/AdministrativeFacade.java [hunk 2]
```diff
@@ -259,44 +265,65 @@ public void updateUserProfilePicture(Map<String, Object> map) throws Exception {
...
+    public boolean isModerator(String jwt , Long communityId)
+    {
+        try {
+            Long userId = authenticationService.getIdFromJwt(jwt);
+            return moderatorService.isModerator(userId,communityId);
+        }
+        catch (Exception e)
+        {
+            return false;
+        }
```
Comment (Ali29Mahmoud) [inline BugByte-backend/src/main/java/com/example/BugByte_backend/facades/AdministrativeFacade.java:300]: not clean structure for the curly braces

File: BugByte-backend/src/main/java/com/example/BugByte_backend/facades/AdministrativeFacade.java [hunk 3]
```diff
@@ -259,44 +265,65 @@ public void updateUserProfilePicture(Map<String, Object> map) throws Exception {
...
+        try {
+            boolean isAdmin = authenticationService.getIsAdminFromJwt(jwt);
+            if (!isAdmin) throw new Exception("user is not an admin");
+
+            return moderatorService.isModeratorByName(userName,communityId);
+        }
+        catch (Exception e)
+        {
+            return false;
+        }
```
Comment (Ali29Mahmoud) [inline BugByte-backend/src/main/java/com/example/BugByte_backend/facades/AdministrativeFacade.java:313]: not a clean structure for the curly braces

File: BugByte-backend/src/main/java/com/example/BugByte_backend/facades/AdministrativeFacade.java [hunk 4]
```diff
@@ -350,4 +377,16 @@ public boolean leaveCommunity(String jwt , String communityName)
...
+    {
+        try {
+            boolean isAdmin = authenticationService.getIsAdminFromJwt(jwt);
+            if (!isAdmin) throw new Exception("user is not an admin");
+            return userService.getUser(username).getIsAdmin();
+        }
+        catch (Exception e)
+        {
+            return false;
+        }
```
Comment (Ali29Mahmoud) [inline BugByte-backend/src/main/java/com/example/BugByte_backend/facades/AdministrativeFacade.java:390]: not a clean structure for the curly braces

File: BugByte-backend/src/main/java/com/example/BugByte_backend/facades/AdministrativeFacade.java [hunk 5]
```diff
@@ -259,44 +265,65 @@ public void updateUserProfilePicture(Map<String, Object> map) throws Exception {
...
+
+            return moderatorService.isModeratorByName(userName,communityId);
+        }
+        catch (Exception e)
+        {
+            return false;
+        }
+    }
+    public boolean removeMember(Long communityId, String memberName,String token)
     {
```
Comment (RowanYoussef) [inline BugByte-backend/src/main/java/com/example/BugByte_backend/facades/AdministrativeFacade.java:316]: we could remove the first method (is admin by id) and  use the username instead

File: BugByte-backend/src/main/java/com/example/BugByte_backend/repositories/CommunityRepository.java [hunk 6]
```diff
@@ -165,8 +166,9 @@ public Long findIdByName(String name) {
 
     @Override
     public Community findCommunityById(Long id) {
-        if( id == null)
+        if( id == null) {
```
Comment (Ali29Mahmoud) [inline BugByte-backend/src/main/java/com/example/BugByte_backend/repositories/CommunityRepository.java:169]: remove the space before id (id == null)

File: BugByte-backend/src/main/java/com/example/BugByte_backend/repositories/CommunityRepository.java [hunk 7]
```diff
@@ -225,23 +227,23 @@ public boolean updateCommunityName(Long communityId, String newName) {
...
-            throw new IllegalArgumentException("communityId is null");
-        }
-        int rows = jdbcTemplate.update(SQL_REMOVE_COMMUNITY_MODERATORS, communityId);
-    }
+//    private void removeCommunityModerators(Long communityId) {
+//        if (communityId == null) {
+//            throw new IllegalArgumentException("communityId is null");
+//        }
+//        int rows = jdbcTemplate.update(SQL_REMOVE_COMMUNITY_MODERATORS, communityId);
+//    }
```
Comment (RowanYoussef) [inline BugByte-backend/src/main/java/com/example/BugByte_backend/repositories/CommunityRepository.java:247]: unnecessary comments

File: BugByte-backend/src/main/java/com/example/BugByte_backend/repositories/CommunityRepository.java [hunk 8]
```diff
@@ -344,10 +346,17 @@ public List<User> findModeratorsByCommunityId(Long communityId) {
...
         return moderators;
     }
+    public boolean  leaveCommunity(String communityName , Long memberId)
+    {
+        if(memberId==null || communityName==null)
+            throw new NullPointerException("memberId or communityId is null");
+        int rows = jdbcTemplate.update(SQL_LEAVE_COMMUNITY, memberId, communityName);
+        return rows == 1;
+    }
+
```
Comment (Ali29Mahmoud) [inline BugByte-backend/src/main/java/com/example/BugByte_backend/repositories/CommunityRepository.java:359]: change the curly braces  for this structure 

public boolean function( ) {
}

File: BugByte-backend/src/main/java/com/example/BugByte_backend/repositories/ModeratorRepository.java [hunk 9]
```diff
@@ -50,5 +56,23 @@ public Boolean removeModerator(Long userId, Long communityId) {
...
+    }
+    public boolean isModeratorByName(String userName, Long communityId)
+    {
+        if (userName == null || communityId == null) {
+            throw new NullPointerException("member Id or community Id is null");
+        }
+        int count = jdbcTemplate.queryForObject(SQL_IS_MODERATOR_BY_NAME, Integer.class, userName, communityId);
+        return count==1;
+
+    }
```
Comment (Ali29Mahmoud) [inline BugByte-backend/src/main/java/com/example/BugByte_backend/repositories/ModeratorRepository.java:76]: fix the curly braces structure of each function

File: BugByte-backend/src/main/java/com/example/BugByte_backend/services/CommunityService.java [hunk 10]
```diff
@@ -231,4 +157,76 @@ public  boolean leaveCommunity(String communityName ,Long memberId)
...
+//        if (community == null) {
+//            throw new IllegalArgumentException("Community not cached: " + communityName);
+//        }
+//        return community;
+//    }
+//    public List<User> getCommunityUsers(Long communityId) {
+//        try {
+//            return communityRepository.getCommunityMembers(communityId);
+//        } catch (Exception e) {
+//            throw e;
```
Comment (Ali29Mahmoud) [inline BugByte-backend/src/main/java/com/example/BugByte_backend/services/CommunityService.java:229]: would we keep this commented part

Comment (habibatarek26) [reply BugByte-backend/src/main/java/com/example/BugByte_backend/services/CommunityService.java:229]: Yes ,I need it but not for now.

**Review by Ali29Mahmoud — COMMENTED** [review]
we need to fix some structures in the code 
the logic should be working as we expecting

**Pull request merged**